src/ethereum_test_tools/code/yul.py

```diff
@@ -4,99 +4,107 @@
 
 
-def get_evm_version_from_fork(fork: Fork | None):
-    """
-    Get the solc evm version corresponding to `fork`.
+class Solc:
+    """Solc compiler."""
 
-    Args
-    ----
-        fork (Fork): The fork to retrieve the corresponding evm version for.
+    binary: Path
 
-    Returns
-    -------
-        str: The name of evm version as required by solc's --evm-version.
-    """
-    if not fork:
-        return None
-    return fork.solc_name()
+    def __init__(
+        self,
+        binary: Optional[Path | str] = None,
+    ):
+        if binary is None:
+            which_path = which("solc")
+            if which_path is not None:
+                binary = Path(which_path)
+        if binary is None or not Path(binary).exists():
+            raise Exception(
+                """`solc` binary executable not found, please refer to
+                https://docs.soliditylang.org/en/latest/installing-solidity.html
+                for help downloading and installing `solc`"""
+            )
+        self.binary = Path(binary)
+
+    def run(self, *args: str, input: str | None = None) -> CompletedProcess:
+        """Run solc with the given arguments"""
+        return run(
+            [self.binary, *args],
+            capture_output=True,
+            text=True,
+            input=input,
+        )
+
+    @cached_property
+    def version(self) -> Version:
+        """Return solc's version"""
+        for line in self.run("--version").stdout.splitlines():
+            if match := VERSION_PATTERN.search(line):
+                # Sanitize
+                solc_version_string = match.group(1).replace("g++", "gpp")
+                return Version.parse(solc_version_string)
+        warnings.warn("Unable to determine solc version.")
+        return Version(0)
 
 
-class Yul(SupportsBytes, Sized):
+class Yul(Solc, SupportsBytes, Sized):
     """
     Yul compiler.
     Compiles Yul source code into bytecode.
     """
 
     source: str
-    compiled: Optional[bytes] = None
+    evm_version: str | None
 
     def __init__(
         self,
         source: str,
         fork: Optional[Fork] = None,
         binary: Optional[Path | str] = None,
     ):
+        super().__init__(binary)
         self.source = source
-        self.evm_version = get_evm_version_from_fork(fork)
-        if binary is None:
-            which_path = which("solc")
-            if which_path is not None:
-                binary = Path(which_path)
-        if binary is None or not Path(binary).exists():
-            raise Exception(
-                """`solc` binary executable not found, please refer to
-                https://docs.soliditylang.org/en/latest/installing-solidity.html
-                for help downloading and installing `solc`"""
-            )
-        self.binary = Path(binary)
+        if fork and (fork := get_closest_fork_with_solc_support(fork, self.version)):
+            self.evm_version = fork.solc_name()
```

Comment:
As our pytest plugins currently modify the fork passed to solc, I think we should avoid potentiallly doing it again in `Yul` here:
```suggestion
        if fork:
            self.evm_version = fork.solc_name()
```

And keep the `Yul` class dumb. I.e., the user can provide any fork to it, if the fork is not supported by solc, then compilation will simply fail. 

I mean, I'm not strictly apposed to refactoring the pytest plugin code and having it here instead (although I'd tend to keep it as it is). But I think it should be in one place.

Here's the relevant pytest fixture from the `test_filler` plugin:
https://github.com/ethereum/execution-spec-tests/blob/c6d12eb97944f8383664c40e9e6de1531e69a090/src/pytest_plugins/test_filler/test_filler.py#L352-L377